File: src/main/java/org/gecko/io/AutomatonFileSerializer.java
package org.gecko.io;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;
import org.gecko.exceptions.ModelException;
import org.gecko.model.Automaton;
import org.gecko.model.Condition;
import org.gecko.model.Contract;
import org.gecko.model.Edge;
import org.gecko.model.GeckoModel;
import org.gecko.model.Kind;
import org.gecko.model.Region;
import org.gecko.model.State;
import org.gecko.model.System;
import org.gecko.model.SystemConnection;
import org.gecko.model.Variable;

/**
 * The AutomatonFileSerializer is used to export a project to a sys file. When exporting, it transforms features unique
 * to Gecko, such as regions, kinds and priorities, to be compatible with the sys file format.
 */
public class AutomatonFileSerializer implements FileSerializer {
    private final GeckoModel model;

    private static final String INDENT = "    ";
    private static final String SERIALIZED_CONTRACT_NAME_REGEX = "contract %s";
    private static final String AUTOMATON_SERIALIZATION_AS_SYSTEM_CONTRACT_REGEX
        = SERIALIZED_CONTRACT_NAME_REGEX + " {";
    private static final String SERIALIZED_CONTRACT_REGEX = SERIALIZED_CONTRACT_NAME_REGEX + " := %s ==> %s";
    private static final String SERIALIZED_TRANSITION_REGEX = "%s -> %s :: %s";
    private static final String SERIALIZED_SYSTEM_REGEX = "reactor %s {";
    private static final String SERIALIZED_CONNECTION_REGEX = "%s.%s -> %s.%s";
    private static final String SERIALIZED_STATE_REGEX = "state %s: %s";
    private static final String VARIABLE_ATTRIBUTES_REGEX = " %s: %s";
    private static final String TRUE_CONDITION = "TRUE";
    private static final String SERIALIZED_INPUT = "input";
    private static final String SERIALIZED_OUTPUT = "output";
    private static final String SERIALIZED_STATE = "state";
    private static final int HIGHEST_PRIORITY = 0;

    public AutomatonFileSerializer(GeckoModel model) {
        this.model = model;
    }

    @Override
    public void writeToFile(File file) throws IOException {
        StringJoiner joiner = new StringJoiner(java.lang.System.lineSeparator());
        if (model.getGlobalDefines() != null) {
            joiner.add(model.getGlobalDefines());
            joiner.add("");
        }
        if (model.getGlobalCode() != null) {
            joiner.add(serializeCode(model.getGlobalCode()));
            joiner.add("");
        }
        joiner.add(serializeAutomata(model));
        joiner.add(serializeSystems(model));
        Files.write(file.toPath(), joiner.toString().getBytes(StandardCharsets.UTF_8));
    }

    private String serializeAutomata(GeckoModel model) {
        List<System> relevantSystems =
            model.getAllSystems().stream().filter(system -> !system.getAutomaton().isEmpty()).toList();
        return serializeCollectionWithMapping(relevantSystems, this::serializeAutomaton);
    }

    private String serializeSystems(GeckoModel model) {
        return serializeCollectionWithMapping(model.getAllSystems(), this::serializeSystem);
    }

    private String serializeAutomaton(System system) {
        Automaton automaton = system.getAutomaton();
        StringJoiner joiner = new StringJoiner(java.lang.System.lineSeparator());
        joiner.add(AUTOMATON_SERIALIZATION_AS_SYSTEM_CONTRACT_REGEX.formatted(system.getName()));
        if (!system.getVariables().isEmpty()) {
            joiner.add(serializeIo(system));
            joiner.add("");
        }
        joiner.add(serializeCollectionWithMapping(automaton.getStates(), this::serializeStateContracts, automaton));
        joiner.add("");
        List<Edge> relevantEdges = automaton.getEdges().stream().filter(edge -> edge.getContract() != null).toList();
        joiner.add(serializeCollectionWithMapping(relevantEdges, this::serializeTransition));
        joiner.add("}");
        joiner.add("");
        return joiner.toString();
    }

    private String serializeStateContracts(State state, Automaton automaton) {
        //Edges are used so much here because contracts don't have priorities or kinds and only states can be in regions
        List<Region> relevantRegions = automaton.getRegionsWithState(state);
        List<Edge> edges =
            automaton.getOutgoingEdges(state).stream().filter(edge -> edge.getContract() != null).toList();
        if (edges.isEmpty()) {
            return "";
        }
        //Creating new contracts to not alter the model
        Map<Contract, Contract> newContracts = new HashMap<>();
        for (Edge edge : edges) {
            Contract newContract = applyRegionsToContract(relevantRegions, edge.getContract());
            try {
                applyKindToContract(newContract, edge.getKind());
            } catch (ModelException e) {
                throw new RuntimeException("Failed to apply kind to contract", e);
            }
            newContracts.put(edge.getContract(), newContract);
        }

        //Building the conditions for the priorities
        List<List<Edge>> groupedEdges =
            new ArrayList<>(edges.stream().collect(Collectors.groupingBy(Edge::getPriority)).values());
        Collections.reverse(groupedEdges); //This just makes this easier to think about
        List<Condition> preConditionsByPrio = new ArrayList<>();
        for (List<Edge> edgeGroup : groupedEdges) {
            //OrElseThrow because validity needs to be ensured by model
            Condition newPre = edgeGroup.stream()
                .map(Edge::getContract)
                .map(newContracts::get)
                .map(Contract::getPreCondition)
                .reduce(Condition::and)
                .orElseThrow();
            preConditionsByPrio.add(newPre);
        }
        //and the specific condition for a prio with all conditions with lower prio
        List<Condition> allLowerPrioPreConditions = new ArrayList<>();
        for (int i = 0; i < preConditionsByPrio.size() - 1; i++) {
            allLowerPrioPreConditions.add(
                allLowerPrioPreConditions.stream().reduce(preConditionsByPrio.get(i), Condition::and));
        }

        //applying priorites
        int prioIndex = HIGHEST_PRIORITY;
        for (List<Edge> edgeGroup : groupedEdges) {
            for (Edge edge : edgeGroup) {
                if (prioIndex == HIGHEST_PRIORITY) {
                    continue; //Highest prio doesn't need to be altered
                }
                Contract contractWithPrio = newContracts.get(edge.getContract());
                contractWithPrio.setPreCondition(
                    contractWithPrio.getPreCondition().and(allLowerPrioPreConditions.get(prioIndex - 1).not()));
                newContracts.put(edge.getContract(), contractWithPrio);
            }
            prioIndex++;
        }
        return serializeCollectionWithMapping(newContracts.values(), this::serializeContract);
    }

    private void applyKindToContract(Contract contract, Kind kind) throws ModelException {
        switch (kind) {
            case MISS -> {
                contract.setPreCondition(contract.getPreCondition().not());
                contract.setPostCondition(new Condition(TRUE_CONDITION));
            }
            case FAIL -> {
                contract.setPostCondition(contract.getPostCondition().not());
            }
            case HIT -> {
            }
            default -> throw new IllegalArgumentException("Unknown kind: " + kind);
        }
    }

    private Contract applyRegionsToContract(List<Region> relevantRegions, Contract contract) {
        Contract newContract;
        try {
            newContract = new Contract(0, contract.getName(), contract.getPreCondition(), contract.getPostCondition());
        } catch (ModelException e) {
            throw new RuntimeException("Failed to build contract out of other valid contracts", e);
        }
        if (relevantRegions.isEmpty()) {
            return newContract;
        }
        List<Condition> newConditions = andConditions(relevantRegions);
        newContract.setPreCondition(newConditions.getFirst());
        newContract.setPostCondition(newConditions.get(1));
        return newContract;
    }

    private List<Condition> andConditions(List<Region> regions) {
        Region first = regions.getFirst();

        Condition newPre;
        Condition newPost;
        try {
            newPre = new Condition(first.getPreAndPostCondition().getPreCondition().getCondition());
            newPost = new Condition(first.getPreAndPostCondition().getPostCondition().getCondition());
        } catch (ModelException e) {
            throw new RuntimeException("Failed to build conditions out of other valid conditions", e);
        }

        for (Region region : regions) {
            newPre = newPre.and(region.getPreAndPostCondition().getPreCondition());
            newPre = newPre.and(region.getInvariant());
            newPost = newPost.and(region.getPreAndPostCondition().getPostCondition());
            newPost = newPost.and(region.getInvariant());
        }
        return List.of(newPre, newPost);
    }

    private String serializeContract(Contract contract) {
        return INDENT + SERIALIZED_CONTRACT_REGEX.formatted(contract.getName(), contract.getPreCondition(),
            contract.getPostCondition());
    }

    private String serializeTransition(Edge edge) {
        return INDENT + SERIALIZED_TRANSITION_REGEX
            .formatted(edge.getSource().getName(), edge.getDestination().getName(), edge.getContract().getName());
    }

    private String serializeSystem(System system) {
        StringJoiner joiner = new StringJoiner(java.lang.System.lineSeparator());
        joiner.add(SERIALIZED_SYSTEM_REGEX.formatted(system.getName()));
        if (!system.getVariables().isEmpty()) {
            joiner.add(serializeIo(system));
            joiner.add("");
        }
        if (!system.getChildren().isEmpty()) {
            joiner.add(serializeChildren(system));
            joiner.add("");
        }
        if (system.getAutomaton() != null && !system.getAutomaton().isEmpty()) {
            joiner.add(INDENT + SERIALIZED_CONTRACT_NAME_REGEX.formatted(system.getName()));
            joiner.add("");
        }
        if (!system.getConnections().isEmpty()) {
            joiner.add(serializeConnections(system));
            joiner.add("");
        }
        if (system.getCode() != null) {
            joiner.add(serializeCode(system.getCode()));
        }
        joiner.add("}");
        joiner.add("");
        return joiner.toString();
    }

    private String serializeConnections(System system) {
        return serializeCollectionWithMapping(system.getConnections(), this::serializeConnection, system);
    }

    private String serializeConnection(SystemConnection connection, System parent) {
        String startSystem = serializeSystemReference(parent, connection.getSource());
        String startPort = connection.getSource().getName();
        String endSystem = serializeSystemReference(parent, connection.getDestination());
        String endPort = connection.getDestination().getName();
        return INDENT + SERIALIZED_CONNECTION_REGEX.formatted(startSystem, startPort, endSystem, endPort);
    }

    private String serializeSystemReference(System parent, Variable var) {
        if (parent.getVariables().contains(var)) {
            return AutomatonFileVisitor.SELF_REFERENCE_TOKEN;
        } else {
            return parent.getChildSystemWithVariable(var).getName();
        }
    }

    private String serializeIo(System system) {
        //TODO group variables of the same type into one line
        List<Variable> orderedVariables =
            system.getVariables().stream().sorted(Comparator.comparing(Variable::getVisibility)).toList();
        return serializeCollectionWithMapping(orderedVariables, this::serializeVariable);
    }

    private String serializeChildren(System system) {
        return serializeCollectionWithMapping(system.getChildren(), this::serializeChild);
    }

    private String serializeChild(System system) {
        return INDENT + SERIALIZED_STATE_REGEX.formatted(system.getName(), system.getName());
    }

    private String serializeVariable(Variable variable) {
        String output = "";
        output += INDENT + switch (variable.getVisibility()) {
            case INPUT -> SERIALIZED_INPUT;
            case OUTPUT -> SERIALIZED_OUTPUT;
            case STATE -> SERIALIZED_STATE;
            default -> throw new IllegalArgumentException("Unknown visibility: " + variable.getVisibility());
        };
        output += VARIABLE_ATTRIBUTES_REGEX.formatted(variable.getName(), variable.getType());
        //TODO append variable value
        return output;
    }

    private String serializeCode(String code) {
        return INDENT + AutomatonFileVisitor.CODE_BEGIN + code + AutomatonFileVisitor.CODE_END;
    }

    private <T> String serializeCollectionWithMapping(Collection<T> collection, Function<T, String> mapping) {
        StringJoiner joiner = new StringJoiner(java.lang.System.lineSeparator());
        for (T item : collection) {
            String result = mapping.apply(item);
            if (result == null || result.isBlank()) {
                continue;
            }
            joiner.add(result);
        }
        return joiner.toString();
    }

    private <T, S> String serializeCollectionWithMapping(
        Collection<T> collection, BiFunction<T, S, String> mapping, S additionalArg) {
        StringJoiner joiner = new StringJoiner(java.lang.System.lineSeparator());
        for (T item : collection) {
            String result = mapping.apply(item, additionalArg);
            if (result == null || result.isBlank()) {
                continue;
            }
            joiner.add(result);
        }
        return joiner.toString();
    }
}

File: src/main/kotlin/org/gecko/io/AutomatonFileSerializer.kt
package org.gecko.io

import java.io.PrintWriter
import java.io.Writer
import java.util.*
import org.gecko.exceptions.ModelException
import org.gecko.viewmodel.*

/**
 * The AutomatonFileSerializer is used to export a project to a sys file. When exporting, it
 * transforms features unique to Gecko, such as regions, kinds and priorities, to be compatible with
 * the sys file format.
 */
class AutomatonFileSerializer(val model: GModel) : FileSerializer {
    lateinit var out: PrintWriter

    override fun writeToStream(w: Writer) {
        out = PrintWriter(w)

        if (model.globalDefines.isNotEmpty()) {
            out.write("defines {")
            model.globalDefines.forEach {
                out.format("\t%s : %s := %s\n", it.name, it.type, it.value)
            }
            out.write("}\n\n")
        }

        model.globalCode?.let {
            if (it.isNotBlank()) {
                serializeCode(it)
                out.write("\n")
            }
        }

        serializeAutomata(model)
        serializeSystems(model)
    }

    private fun serializeAutomata(model: GModel) =
        model.allSystems.forEach { serializeAutomaton(it) }

    private fun serializeSystems(model: GModel) = model.allSystems.forEach { serializeSystem(it) }

    private fun serializeAutomaton(system: System) {
        val automaton = system.automaton

        out.format(AUTOMATON_SERIALIZATION_AS_SYSTEM_CONTRACT, system.name)

        if (system.ports.isNotEmpty()) {
            serializeIo(system)
            out.write("\n")
        }
        automaton.states.forEach { state -> this.serializeStateContracts(state, automaton) }
        out.write("\n")
        val relevantEdges = automaton.edges.filter { edge: Edge -> edge.contract != null }
        relevantEdges.forEach { edge -> this.serializeTransition(edge) }
        out.write("}")
        out.write("\n")
    }

    private fun serializeStateContracts(state: State, automaton: Automaton) {
        // Edges are used so much here because contracts don't have priorities or kinds and only
        // states can be in regions
        val relevantRegions = automaton.getRegionsWithState(state)
        val edges = automaton.getOutgoingEdges(state).filter { it.contract != null }
        if (edges.isEmpty()) {
            return
        }
        // Creating new contracts to not alter the model
        val newContracts: MutableMap<Edge, Contract> = HashMap()
        for (edge in edges) {
            val newContract = applyRegionsToContract(relevantRegions, edge.contract!!)
            try {
                applyKindToContract(newContract, edge.kind)
                newContract.name = getContractName(edge)
            } catch (e: ModelException) {
                throw RuntimeException("Failed to apply kind to contract", e)
            }
            newContracts[edge] = newContract
        }

        // Building the conditions for the priorities
        val groupedEdges = edges.groupBy { it!!.priority }.values.reversed()
        val preConditionsByPrio: MutableList<Condition> = arrayListOf()
        for (edgeGroup in groupedEdges) {
            // OrElseThrow because validity needs to be ensured by model
            val newPre =
                edgeGroup
                    .map { key -> newContracts[key]!! }
                    .map { obj -> obj.preCondition }
                    .reduce { obj, other -> Condition("$obj & $other") }
            preConditionsByPrio.add(newPre)
        }
        // and the specific condition for a prio with all conditions with lower prio
        val allLowerPrioPreConditions = arrayListOf<Condition>()
        for (i in 0 until preConditionsByPrio.size - 1) {
            allLowerPrioPreConditions.add(
                allLowerPrioPreConditions.reduce { obj, other -> obj.and(other) }
            )
        }

        // applying priorites
        for ((prioIndex, edgeGroup) in groupedEdges.withIndex()) {
            for (edge in edgeGroup) {
                if (prioIndex == 0) {
                    continue // Highest prio doesn't need to be altered
                }
                val contractWithPrio = newContracts[edge]!!
                contractWithPrio.preCondition.value =
                    contractWithPrio.preCondition
                        .and(allLowerPrioPreConditions[prioIndex - 1].not())
                        .value
                newContracts[edge] = contractWithPrio
            }
        }
        newContracts.values.forEach { this.serializeContract(it) }
    }

    @Throws(ModelException::class)
    fun applyKindToContract(contract: Contract, kind: Kind) {
        when (kind) {
            Kind.MISS -> {
                contract.preCondition.value = contract.preCondition.not().value
                contract.postCondition.value = Condition("true").value
            }
            Kind.FAIL -> contract.postCondition.value = contract.postCondition.not().value
            Kind.HIT -> {}
        }
    }

    private fun applyRegionsToContract(
        relevantRegions: List<Region>,
        contract: Contract
    ): Contract {
        val newContract: Contract
        try {
            newContract = Contract(contract.name, contract.preCondition, contract.postCondition)
        } catch (e: ModelException) {
            throw RuntimeException("Failed to build contract out of other valid contracts", e)
        }
        if (relevantRegions.isEmpty()) {
            return newContract
        }
        val (pre, post) = andConditions(relevantRegions)
        newContract.preCondition = pre and newContract.preCondition
        newContract.postCondition = post and newContract.postCondition
        return newContract
    }

    private fun andConditions(regions: List<Region>): Pair<Condition, Condition> {
        val c = regions.first()
        val first = c.contract

        var newPre: Condition
        var newPost: Condition
        try {
            newPre = first.preCondition
            newPost = first.postCondition
        } catch (e: ModelException) {
            throw RuntimeException("Failed to build conditions out of other valid conditions", e)
        }
        newPre = newPre.and(c.invariant)
        newPost = newPost.and(c.invariant)

        for (i in 1 until regions.size) {
            val region = regions[i]
            val con = region.contract
            newPre = newPre.and(con.preCondition)
            newPre = newPre.and(region.invariant)
            newPost = newPost.and(con.postCondition)
            newPost = newPost.and(region.invariant)
        }
        return newPre to newPost
    }

    private fun serializeContract(contract: Contract) {
        out.write(INDENT)
        out.format(
            SERIALIZED_CONTRACT,
            contract.name,
            contract.preCondition,
            contract.postCondition
        )
    }

    private fun serializeTransition(edge: Edge) {
        out.write(INDENT)
        out.format(
            SERIALIZED_TRANSITION,
            edge.source.name,
            edge.destination.name,
            getContractName(edge)
        )
    }

    private fun serializeSystem(system: System) {
        out.format(SERIALIZED_SYSTEM, system.name)

        if (system.ports.isNotEmpty()) {
            serializeIo(system)
            out.write("\n")
        }
        if (system.subSystems.isNotEmpty()) {
            serializeChildren(system)
            out.write("\n")
        }
        out.write(INDENT)
        out.format(SERIALIZED_CONTRACT_NAME, system.name)
        out.write("\n")
        if (system.connections.isNotEmpty()) {
            serializeConnections(system)
            out.write("\n")
        }
        serializeCode(system.code)
        out.write("}")
        out.write("\n")
    }

    private fun serializeConnections(system: System) =
        system.connections.forEach { this.serializeConnection(it, system) }

    private fun serializeConnection(connection: SystemConnection, parent: System) {
        val startSystem = serializeSystemReference(parent, connection.source!!)
        val startPort = connection.source?.name
        val endSystem = serializeSystemReference(parent, connection.destination!!)
        val endPort = connection.destination?.name
        out.write(INDENT)
        out.format(SERIALIZED_CONNECTION, startSystem, startPort, endSystem, endPort)
    }

    private fun serializeSystemReference(parent: System?, v: Port): String {
        return if (parent!!.ports.contains(v)) {
            AutomatonFileVisitor.SELF_REFERENCE_TOKEN
        } else {
            parent.getChildSystemWithVariable(v)!!.name
        }
    }

    private fun serializeIo(system: System) {
        val orderedVariables = system.ports.sorted(Comparator.comparing { it.visibility })
        orderedVariables.forEach { this.serializeVariable(it) }
    }

    private fun serializeChildren(system: System) =
        system.subSystems.forEach { this.serializeChild(it) }

    private fun serializeChild(system: System): String =
        String.format(INDENT + SERIALIZED_STATE, system.name, system.name)

    private fun serializeVariable(variable: Port): String {
        var output = ""
        output +=
            INDENT +
                when (variable.visibility) {
                    Visibility.INPUT -> SERIALIZED_INPUT
                    Visibility.OUTPUT -> SERIALIZED_OUTPUT
                    Visibility.STATE -> SERIALIZED_STATE_VISIBILITY
                }
        output += VARIABLE_ATTRIBUTES.format(variable.name, variable.type)
        if (variable.value != null) {
            output += " := " + variable.value
        }
        return output
    }

    private fun serializeCode(code: String) =
        out.write(INDENT + AutomatonFileVisitor.CODE_BEGIN + code + AutomatonFileVisitor.CODE_END)

    private fun getContractName(edge: Edge): String {
        val name = edge.contract?.name
        if (edge.kind == Kind.HIT) {
            return name!!
        }
        return makeNameUnique("%s %s".format(name, edge.kind.name))
    }

    private fun makeNameUnique(baseName: String): String = baseName

    companion object {
        const val INDENT = "    "
        const val SERIALIZED_CONTRACT_NAME = "contract %s"
        const val AUTOMATON_SERIALIZATION_AS_SYSTEM_CONTRACT = "$SERIALIZED_CONTRACT_NAME {"
        const val SERIALIZED_CONTRACT = "$SERIALIZED_CONTRACT_NAME := %s ==> %s"
        const val SERIALIZED_TRANSITION = "%s -> %s :: %s"
        const val SERIALIZED_SYSTEM = "reactor %s {"
        const val SERIALIZED_CONNECTION = "%s.%s -> %s.%s"
        const val SERIALIZED_STATE = "state %s: %s"
        const val VARIABLE_ATTRIBUTES = " %s: %s"
        const val SERIALIZED_INPUT = "input"
        const val SERIALIZED_OUTPUT = "output"
        const val SERIALIZED_STATE_VISIBILITY = "state"
    }
}
